libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -630,6 +630,104 @@ string YulUtilFunctions::overflowCheckedIntExpFunction(
 	});
 }
 
+string YulUtilFunctions::overflowCheckedIntLiteralExpFunction(
+	RationalNumberType const& _baseType,
+	IntegerType const& _exponentType,
+	IntegerType const& _commonType
+)
+{
+	solAssert(!_exponentType.isSigned(), "");
+	solAssert(_baseType.isNegative() == _commonType.isSigned(), "");
+	solAssert(_commonType.numBits() == 256, "");
+
+	// Calculates the upperbound for exponentiation, that is, calculate `b`, such that
+	// _base**b <= _maxValue and _base**(b + 1) > _maxValue
+	auto findExponentUpperbound = [](bigint const _base, bigint const _maxValue) -> unsigned
+	{
+		unsigned first = 0;
+		unsigned last = 255;
+		unsigned middle;
+
+		while (first < last)
+		{
+			middle = (first + last) / 2;
+
+			if (
+				// The condition on msb is a shortcut that avoids computing large powers in
+				// arbitrary precision.
+				boost::multiprecision::msb(_base) * middle <= boost::multiprecision::msb(_maxValue) &&
+				boost::multiprecision::pow(_base, middle) <= _maxValue
+			)
+			{
+				if (boost::multiprecision::pow(_base, middle + 1) > _maxValue)
+					return middle;
+				else
+					first = middle + 1;
+			}
+			else
+				last = middle;
+		}
+
+		return last;
+	};
+
+
+	u256 baseValue = _baseType.literalValue(nullptr);
+	unsigned exponentUpperbound;
+
+	if (_baseType.isNegative())
+	{
+		s256 signedValue = u2s(baseValue);
```

Comment:
Why not use bigint for everything?

Reply:
`RationalNumberType::literalValue()` returns an `u256` and negative numbers in two's complement. Maybe we can do the right conversion and do everything in bigint. I'll check.

Reply:
I checked this, for the literal in the whisker's template, we need two's complement. So if we do everything in bigint, we'll need two conversions, from `RationalType::literalValue()` (which is u256 in two's complement) to `bigint` and again back to two's complement. I think it's not worth it.

Reply:
I still find `signedValue` and `baseValue` very confusing. What about having a `bigint baseValue` being the actual value and in addition a `u256 baseValueTwosComplement` variable that holds the two's complement? On the other hand, the only place where you need the two's complement is in the whiskers template, which could just be

`(baseValue < 0 ? (bigint(1) << 256) - baseValue : baseValue).str()